### PR TITLE
[WIP] HTTP2 support

### LINKED ISSUE #167: HTTP2 support
[Issuehunt badges](https://issuehunt.io/r/sindresorhus/got/issues/167)

Hi!

Can you consider adding `HTTP2` support to GOT?
Currently the most used client library is this: https://github.com/molnarg/node-http2

Thank you!

> There is a $200.00 open bounty on this issue. Add more on [Issuehunt](https://issuehunt.io/r/sindresorhus/got/issues/167).
> - Checkout the [Issuehunt explorer](https://issuehunt.io/r/sindresorhus/got/) to discover more funded issues.
> - Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on Issuehunt to raise funds.

### PR DESCRIPTION
**Pull request by szmarczak**
Fixes #167 

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [x] If it's a new feature, I have included documentation updates.

### CODE REVIEW

File: readme.md [hunk 1]
```diff
@@ -1188,27 +1189,27 @@ const custom = got.extend({
 
 **Tip:** Need to merge some instances into a single one? Check out [`got.mergeInstances()`](advanced-creation.md#merging-instances).
 
-### Experimental HTTP2 support
+### HTTP2 support
 
-Got provides an experimental support for HTTP2 using the [`http2-wrapper`](https://github.com/szmarczak/http2-wrapper) package:
+Got supports HTTP2 via the [`http2-wrapper`](https://github.com/szmarczak/http2-wrapper) package.<br>
+To make this work, you need to use Node v10 or higher.
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
To make this work, you need to use Node.js 10 or higher.
```

File: readme.md [hunk 2]
```diff
@@ -1188,27 +1189,27 @@ const custom = got.extend({
...
+
+**Note:** using custom `request` function will disable HTTP2 support.
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
**Note:** Using the `request` option will disable HTTP2 support.
```

File: readme.md [hunk 3]
```diff
@@ -1188,27 +1189,27 @@ const custom = got.extend({
...
 const got = require('got');
-const {request} = require('http2-wrapper');
-
-const h2got = got.extend({request});
 
 (async () => {
-	const {body} = await h2got('https://nghttp2.org/httpbin/headers');
-	console.log(body);
+	const {headers} = await got('https://nghttp2.org/httpbin/anything');
+	console.log(headers.via); // => 2 nghttpx
```
Comment (sindresorhus) [inline readme.md]:
```suggestion
	console.log(headers.via);
	//=> '2 nghttpx'
```

File: source/request-as-event-emitter.ts [hunk 4]
```diff
@@ -71,16 +72,17 @@ export default (options: NormalizedOptions, input?: TransformStream) => {
...
-		} else {
-			requestFn = options.protocol === 'https:' ? https.request : http.request;
-		}
+		let requestFn: RequestFunction = options.request;
 
-		if (agents) {
-			const protocolName = options.protocol === 'https:' ? 'https' : 'http';
-			options.agent = (agents as AgentByProtocol)[protocolName] || options.agent;
+		if (requestFn) {
+			// Custom `request` function, we need to prepare `options.agent` on our own.
```
Comment (sindresorhus) [inline source/request-as-event-emitter.ts]:
```suggestion
			// With a custom `request` function, we need to prepare `options.agent` on our own.
```

File: test/https.ts [hunk 5]
```diff
@@ -20,3 +21,8 @@ test('https request with ca', withServer, async (t, server, got) => {
 	});
 	t.is(body, 'ok');
 });
+
+test('http2 works', async t => {
```
Comment (sindresorhus) [inline test/https.ts]: Would be good with some more HTTP2 tests. Should be in a separate `http2.js` test file.

File: readme.md [hunk 6]
```diff
@@ -397,9 +398,9 @@ Default: `false`
 ###### request
 
 Type: `Function`<br>
-Default: `http.request` `https.request` *(Depending on the protocol)*
+Default: `http.request` `https.request` `http2wrapper.request` *(Depending on the protocol)*
 
-Custom request function. The main purpose of this is to [support HTTP2 using a wrapper](#experimental-http2-support).
+Custom request function. The main purpose of this is to [support HTTP2 using a wrapper](#http2-support).
```
Comment (sindresorhus) [inline readme.md]: > The main purpose of this is to [support HTTP2 using a wrapper](#http2-support).

This is misleading as it's no built-in. I would just drop that sentence or document a different use-case.

File: readme.md [hunk 7]
```diff
@@ -397,9 +398,9 @@ Default: `false`
 ###### request
 
 Type: `Function`<br>
-Default: `http.request` `https.request` *(Depending on the protocol)*
+Default: `http.request` `https.request` `http2wrapper.request` *(Depending on the protocol)*
```
Comment (sindresorhus) [inline readme.md]: Can you linkify `http2wrapper.request` to your package?

**Comment by szmarczak** [comment]
@sindresorhus Maybe let's make all the tests use a HTTP2 server instead of HTTP1? That way we would be sure that `http2-wrapper` works as expected.

**Comment by sindresorhus** [comment]
> Maybe let's make all the tests use a HTTP2 server instead of HTTP1? That way we would be sure that http2-wrapper works as expected.

Ideally, we would run all tests on both a HTTP1 and HTTP2 server to ensure everything works in both modes.

**Comment by szmarczak** [comment]
> Ideally, we would run all tests on both a HTTP1 and HTTP2 server to ensure everything works in both modes.

So... Use an environment variable? And then run AVA twice?

File: source/request-as-event-emitter.ts [hunk 8]
```diff
@@ -207,6 +208,35 @@ export default (options: NormalizedOptions): RequestAsEventEmitter => {
...
+				// @ts-ignore Type URL is not assignable to type URL
+				const result = request(url, options, callback);
+
+				if (is.promise(result)) {
+					const emitter = new EventEmitter();
+
+					// @ts-ignore `cacheable-request` doesn't support async request function
+					result.once = emitter.once.bind(emitter);
+
+					// @ts-ignore This is a TS bug, because `result` is `Promise<http.ClientRequest>`
```
Comment (szmarczak) [inline source/request-as-event-emitter.ts]: It complains that `request` is `unknown`

File: source/request-as-event-emitter.ts [hunk 9]
```diff
@@ -207,6 +208,35 @@ export default (options: NormalizedOptions): RequestAsEventEmitter => {
...
+					// eslint-disable-next-line promise/prefer-await-to-then
+					result.then((request: http.ClientRequest) => {
```
Comment (szmarczak) [inline source/request-as-event-emitter.ts]: `return result.then(...);`

File: source/request-as-event-emitter.ts [hunk 10]
```diff
@@ -218,12 +248,18 @@ export default (options: NormalizedOptions): RequestAsEventEmitter => {
 				}
 			});
 
-			cacheRequest.once('request', handleRequest);
+			cacheRequest.once('request', async (request: Promisable<http.ClientRequest>) => {
```
Comment (szmarczak) [inline source/request-as-event-emitter.ts]: `Promisable` -> `ClientRequest | Promise<ClientRequest>`

**Comment by sindresorhus** [comment]
> So... Use an environment variable? And then run AVA twice?

Yeah

**Comment by szmarczak** [comment]
Closing in favor of #1051

**Pull request closed without merging**